src/channels.ml

```diff
@@ -23,7 +23,18 @@ external _sendfile: Unix.file_descr -> Unix.file_descr -> int64 -> int64 = "stub
 let _sendfile from_fd to_fd len =
   let from_fd = Lwt_unix.unix_file_descr from_fd in
   let to_fd = Lwt_unix.unix_file_descr to_fd in
-  detach (_sendfile from_fd to_fd) len
+  let max_attempts = 20 in
+  let rec loop remaining_attempts =
+    Lwt.catch
+      (fun () -> detach (_sendfile from_fd to_fd) len)
+      (function
+        | Unix.(Unix_error (EAGAIN, _, _)) as e when remaining_attempts > 0 ->
```

Comment:
FWIW the right way to retry syscalls in Lwt is to use Lwt_unix.wrap_syscall, there is no guarantee that the socket would be available by just retrying after a certain amount of time, wrap_syscall uses Lwt's event mechanism to wake up the next time the socket is available for read or write: https://ocsigen.org/lwt/3.1.0/api/Lwt_unix

Reply:
Can you update the PR to do the right thing with `wrap_syscall` or you think it is a too big and risky change for now?

Reply:
I'll open a ticket, using wrap_syscall would require the socket to be in non-blocking mode, which  according to the comment should be but the implementation does the opposite (sets it to blocking mode): https://github.com/xapi-project/vhd-tool/blob/master/src/channels.ml#L31-L40
Also `sendfile` can potentially send gigabytes of data and run for a very long time, so we need to keep using Lwt_preemptive.detach, and that doesn't work with Lwt_unix.wrap_syscall.

I think the reason why we see EAGAIN on qemu-datapath branch is because our source is now /dev/nbd0, which maybe acts more like a socket than an actual file, but not sure it supports non-blocking I/O properly. According to `sendfile`'s manpage you get EAGAIN when your output is in non-blocking mode and you would block, it seems here our input would block.